Try resetting the GPU process if rendering commands time out.

mattwoodrow opened the pull request:
#### 7f23ebca08f79530f3c7671579a37643a4e3812b
<pre>
Try resetting the GPU process if rendering commands time out.
<a href="https://bugs.webkit.org/show_bug.cgi?id=274934">https://bugs.webkit.org/show_bug.cgi?id=274934</a>
&lt;<a href="https://rdar.apple.com/125143181">rdar://125143181</a>&gt;

Reviewed by NOBODY (OOPS!).

We have a few issues where rendering commands start timing out, and aren&apos;t being processed
by the GPU.

Ideally we&apos;ll fix all the root cause issues, but in the interim it is probably worth using a
restart of the GPUP to get out of this state

This is restricted to only once per WebProcess, to limit the capabilities of an untrusted
WebProcess.

* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::terminateGPUProcess):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/UIProcess/WebProcessProxy.messages.in:
* Source/WebKit/WebProcess/GPU/GPUProcessConnection.cpp:
(WebKit::GPUProcessConnection::notifyMessageDidTimeout):
* Source/WebKit/WebProcess/GPU/GPUProcessConnection.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.cpp:
(WebKit::RemoteDisplayListRecorderProxy::send):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::~RemoteRenderingBackendProxy):
(WebKit::RemoteRenderingBackendProxy::ensureGPUProcessConnection):
(WebKit::RemoteRenderingBackendProxy::send):
(WebKit::RemoteRenderingBackendProxy::sendSync):
(WebKit::RemoteRenderingBackendProxy::disconnectGPUProcess):
(WebKit::RemoteRenderingBackendProxy::notifyMessageDidTimeout):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::terminateGPUProcess):
* Source/WebKit/WebProcess/WebProcess.h:
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7f23ebca08f79530f3c7671579a37643a4e3812b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/54174 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/33562 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/6712 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/57450 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/4898 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/41084 "Built successfully") | [❌ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/4871 "Failed to checkout and rebase branch from PR 29342") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/43873 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/3273 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/56270 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/31807 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/46909 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/25020 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/28613 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/3047 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/50319 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/4436 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/59043 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/29380 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/4587 "Build was cancelled. Recent messages:OS: Sonoma (14.4.1), Xcode: 15.3; Running apply-patch; Checked out pull request; Skipped layout-tests; layout-tests (exception)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/51292 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/30549 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/47021 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/50648 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/31516 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/30329 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/UIProcess/WebProcessPool.cpp:
```diff
@@ -567,6 +567,12 @@ void WebProcessPool::createGPUProcessConnection(WebProcessProxy& webProcessProxy
+void WebProcessPool::resetGPUProcess()
+{
+    if (auto* process = gpuProcess())
+        process->terminateForTesting();
```
Why is this calling terminateForTesting() instead of terminate()? terminateForTesting() calls `gpuProcessExited(ProcessTerminationReason::IdleExit)` which doesn't seem right since this is not an IDLE exit.

cdumez commented on Source/WebKit/UIProcess/WebProcessPool.cpp:
```diff
@@ -567,6 +567,12 @@ void WebProcessPool::createGPUProcessConnection(WebProcessProxy& webProcessProxy
+
+void WebProcessPool::resetGPUProcess()
+{
+    if (auto* process = gpuProcess())
```
RefPtr

cdumez commented on Source/WebKit/UIProcess/WebProcessPool.cpp:
```diff
@@ -567,6 +567,12 @@ void WebProcessPool::createGPUProcessConnection(WebProcessProxy& webProcessProxy
     ensureProtectedGPUProcess()->createGPUProcessConnection(webProcessProxy, WTFMove(connectionIdentifier), WTFMove(parameters));
 }
+
+void WebProcessPool::resetGPUProcess()
```
Unclear why we need a new function on the WebProcessPool. Why can't the caller just call terminate() on the GPUProcessProxy directly?

cdumez commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -1073,6 +1073,15 @@ void WebProcessProxy::createGPUProcessConnection(IPC::Connection::Handle&& conne
 
+void WebProcessProxy::resetGPUProcess()
+{
+    WEBPROCESSPROXY_RELEASE_LOG(Process, "resetGPUProcess: webPage=%p, pageProxyID=%" PRIu64 ", webPageID=%" PRIu64 " hasReset=%d", &webPage, webPage.identifier().toUInt64(), webPage.webPageID().toUInt64(), m_hasResetGPUProcess);
```
This should at the very least be a RELEASE_LOG_ERROR, maybe even a RELEASE_LOG_FAULT.

cdumez commented on Source/WebKit/UIProcess/WebProcessProxy.messages.in:
```diff
@@ -35,6 +35,7 @@ messages -> WebProcessProxy LegacyReceiver {
 
 #if ENABLE(GPU_PROCESS)
     CreateGPUProcessConnection(IPC::ConnectionHandle connectionHandle, struct WebKit::GPUProcessConnectionParameters parameters) AllowedWhenWaitingForSyncReply
+    ResetGPUProcess()
```
"Reset" could mean many things. I think we should be clear and call this "TerminateGPUProcess".

Now I'll note that it is fishy to add an IPC that allows the WebContent process to kill the GPUProcess, given that the GPUProcess is shared by all processes. I see that you mitigated that by allowing the WebProcess to only kill the GPUProcess once.

It still feels like a pretty big hack to paper over a real bug somewhere else.

mattwoodrow replied:
It is entirely a big hack to paper over a real bug :) I did mention that in the commit message too.

I don't think we should stop trying to fix the real bug, but this is hopefully a user experience improvement in the interim.

cdumez commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -1382,6 +1382,11 @@ GPUProcessConnection& WebProcess::ensureGPUProcessConnection()
     return *m_gpuProcessConnection;
 }
 
+void WebProcess::resetGPUProcess()
```
"reset" is too vague.

cdumez commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -1073,6 +1073,15 @@ void WebProcessProxy::createGPUProcessConnection(IPC::Connection::Handle&& conne
+    WEBPROCESSPROXY_RELEASE_LOG(Process, "resetGPUProcess: webPage=%p, pageProxyID=%" PRIu64 ", webPageID=%" PRIu64 " hasReset=%d", &webPage, webPage.identifier().toUInt64(), webPage.webPageID().toUInt64(), m_hasResetGPUProcess);
+    if (!m_hasResetGPUProcess) {
+        m_hasResetGPUProcess = true;
+        protectedProcessPool()->resetGPUProcess();
```
If we think that the issue is happening because the GPUProcess is hung, we may be able to adopt another approach. Instead of asking us to kill the GPUProcess, the WebProcess could ask the UIProcess to check if the GPUProcess is responsive (we already have logic to check that). Then if the UIProcess detects the GPUProcess is unresponsive, it would kill it (We likely already have logic to kill on unresponsive GPUProcess detection).

Then it would feel less hacky to me and there would be no need to only allow it once per WebProcess. This means we could recover from bad behavior more than once, per lifetime of a WebProcess.

Note that a WebProcess may live for a long time due to the process cache and may get reused for plenty of loads.

mattwoodrow replied:
The issue is not that the GPUProcess is truly hung, but instead that it's idle and not doing any work, while the WebProcess is blocked waiting for it.

cdumez replied:
> The issue is not that the GPUProcess is truly hung, but instead that it's idle and not doing any work
Are we sure that the GPUProcess is not suspended?

cdumez replied:
I just looked at the spindump and it doesn't appear the GPUProcess is suspended. It does look idle though, and not hung, weird. Must be some kind of IPC issue?

mattwoodrow replied:
Yeah, almost certainly some sort of race/logic bug in IPC.

kkinnunen-apple reviewed (commented):
I mentioned this in passing in the Slack messages:

When a hang or real connection loss happens, it will likely happen all N distinct producer engines: RRB, GL, WebGPU. They also may happen in very asynchronous manner.

We should not have N GPUP restarts. We should have 1 restart for all the same generation of hangs.

When establishing connection, we need to capture a weak ptr to the GPUProcessConnection.

When the engine detects the hang, it should let the GPUProcessConnection know. This should be happening with a task posted to main thread or executed directly inline  in case the current code runs in main thread.

The engines like RRB need to ensure that once the task has been posted, future sends don't re-send these request.

Upon executing the task to notify that gpup is timing out:

If the GPUProcessConnection does not exist anymore, it means the issue has been handled.

GPUProcessConnection should contain the logic to let the UI process know something is off with the GPUP connection.

mattwoodrow committed:
Try resetting the GPU process if rendering commands time out.
https://bugs.webkit.org/show_bug.cgi?id=274934
<rdar://125143181>

Reviewed by NOBODY (OOPS!).

We have a few issues where rendering commands start timing out, and aren't being processed
by the GPU.

Ideally we'll fix all the root cause issues, but in the interim it is probably worth using a
restart of the GPUP to get out of this state

This is restricted to only once per WebProcess, to limit the capabilities of an untrusted
WebProcess.

* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::terminateGPUProcess):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/UIProcess/WebProcessProxy.messages.in:
* Source/WebKit/WebProcess/GPU/GPUProcessConnection.cpp:
(WebKit::GPUProcessConnection::notifyMessageDidTimeout):
* Source/WebKit/WebProcess/GPU/GPUProcessConnection.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.cpp:
(WebKit::RemoteDisplayListRecorderProxy::send):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::~RemoteRenderingBackendProxy):
(WebKit::RemoteRenderingBackendProxy::ensureGPUProcessConnection):
(WebKit::RemoteRenderingBackendProxy::send):
(WebKit::RemoteRenderingBackendProxy::sendSync):
(WebKit::RemoteRenderingBackendProxy::disconnectGPUProcess):
(WebKit::RemoteRenderingBackendProxy::notifyMessageDidTimeout):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::terminateGPUProcess):
* Source/WebKit/WebProcess/WebProcess.h:

kkinnunen-apple reviewed (commented):
Were you able to test this somehow? How do we Keep it working?

kkinnunen-apple commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -1090,6 +1090,16 @@ void WebProcessProxy::createGPUProcessConnection(IPC::Connection::Handle&& conne
+{
+    WEBPROCESSPROXY_RELEASE_LOG_ERROR(Process, "terminateGPUProcess: hasTerminate=%d", m_hasTerminatedGPUProcess);
+    if (!m_hasTerminatedGPUProcess) {
+        m_hasTerminatedGPUProcess = true;
```
Do we ever flip this flag back?

Doesn't the weak ptr in WP offer enough protection?

mattwoodrow replied:
The concern is that a compromised WebProcess could repeatedly request GPU process restarts, and cause a denial of service.

So, no, we don't :) Switching from a bool to a counter with a low limit is an option if we think it's worthwhile.

kkinnunen-apple replied:
The denial of service from compromised web process is not in the threat model today.. We don't test for it.

The compromised web process is able to  choose a crashe in GPUP whenever it wants even without this patch.

kkinnunen-apple commented on Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
```diff
@@ -156,6 +164,16 @@ void RemoteRenderingBackendProxy::disconnectGPUProcess()
+
+void RemoteRenderingBackendProxy::notifyMessageDidTimeout()
+{
+    ensureOnMainRunLoop([weakGPUProcessConnection = m_gpuProcessConnection]() {
```
I think you need to move the weak ptr out, so that only one message is sent.
Also the connection probably would need to be invalidated now, e.g. you'd need to run the disconnectGPUProcess phase?

mattwoodrow replied:
The intention was for this to use the same member as you're using in [#29173 ](https://github.com/WebKit/WebKit/pull/29173), so we can't move it out as its used later on.

I don't think we need to invalidate it, the expectation is that the GPU process will 'crash' in the near future and we'd take the usual restart path at that point. The calling code here will just attempt to carry on (and probably immediately block in the next message send attempt).

kkinnunen-apple replied:
> as its used later on.

But for what is it used later on?
When the message send to GPUP times out, the GPUP is defunct.It cannot be trusted and there's nothing to be done with it.

Also we cannot let anybody send any more messages, as the buffer is in unknown state (it will assert after timeout)..

Yes, at the moment invalidation would actually be problematic, as that will disconnect the didClose handler that this commit depends on..

kkinnunen-apple commented on Source/WebKit/WebProcess/GPU/GPUProcessConnection.h:
```diff
@@ -101,6 +101,8 @@ class GPUProcessConnection : public ThreadSafeRefCountedAndCanMakeThreadSafeWeak
     void setMediaEnvironment(WebCore::PageIdentifier, const String&);
 #endif
 
+    void notifyMessageDidTimeout();
```
Not sure this "message did timeout" is the right concept here. It's not really related to all messages.
Maybe terminateProcess() would sound better?

I'm not very keen on the flagging either, eg..the m_hasTimeout. I'd very much expect that the connection object would go away directly at the call, similar to after connection closing. I do understand it seems to be easier to implement that you'd do something that closes the connection, and then let the connection events handle the actual teardown..

mattwoodrow replied:
At some point we need to switch from the symptom (message timed out), to the requested action (terminate the GPU process). I don't have strong opinions on where that boundary is really, happy to make it not-here if you prefer.

I'm worried that making the connection object go away immediately adds a new tear down path that isn't used anywhere else. We have existing code for making sure we recreate everything in response to a crash, so making use of that seems beneficial.

kkinnunen-apple replied:
Yeah, I think the symptom doesn't belong here anymore..
